File: Plant.cpp
#include "CONSTANTS.h"

#include "World.h"
#include "Plant.h"
#include <stdlib.h>

Plant::Plant(World &world, const Point &position, const char &symbol, const int &strength, const std::string &color)
  : Organism(world, position, symbol, strength, PLANT_INITIATIVE, color), 
  spreadProbability(SPREAD_PROBABILITY) {}

Plant::~Plant() {}

void Plant::Action() {
  if(CanAction()){
    this->TryToSpread();
  }
}

bool Plant::Collision(Organism* attackerOrganism) {
  this->Die();
  return false;
}

double Plant::GetSpreadProbability() const {
  return spreadProbability;
}

void Plant::TryToSpread() {
  if (rand() % 100 < spreadProbability * 100) {
    Point newPosition = this->world.GetRandomPositionAround(this->GetPosition(), true, 1);
    if (newPosition != this->GetPosition()) {
      Organism* newPlant = this->Clone(newPosition);
      this->world.AddOrganism(newPlant);
    }
  }
}

File: Java/Plant.java
import java.awt.*;

public abstract class Plant extends Organism {
  private final double spreadChance;

  Plant(World world, Point position, char symbol, int strength, Color color, String name) {
    super(world, position, symbol, strength, Constants.PLANT_INITIATIVE, color, name);
    this.spreadChance = Constants.SPREAD_PROBABILITY;
  }

  private void TryToSpread() {
    Point newPosition = world.GetRandomPositionAround(this.GetPosition(), true, 1);
    if (newPosition != null) {
      Organism newOrganism = this.Clone(newPosition);
      world.AddOrganism(newOrganism);
    }
  }

  @Override
  public void Action() {
    if(CanAction()) {
      if (Math.random() < spreadChance) {
        this.TryToSpread();
      }
    }
  }

  @Override
  public boolean Collision(Organism other) {
    this.Die();
    this.world.messageManager.AddPlantEatingMessage(other.GetName(), this.GetName());
    return false;
  }
}
